pyqt6: init at 6.3.1

milahu opened the pull request:
continue #175510 by @nrdxp

**Description of changes**

add qt6 (#169296) bindings for python.
we already have pyside6 as qt6-for-python
but pyqt6 and pyside6 have different APIs, so we need both

fix some build tools, to build pyqt6 with cmake.
originally, pyqt6 is built with qmake, but that is broken in qt6 (cannot find qt modules)
fixing qmake is less useful, as since qt6, qmake is deprecated in favor of cmake

**upstream issues**

* qmake2cmake
  * [QTBUG-104194](https://bugreports.qt.io/browse/QTBUG-104194) to implement target install commands
  * [QTBUG-104195](https://bugreports.qt.io/browse/QTBUG-104195) to fix parsing of literal strings. simple workaround: wrap string in quotes
* sip
  * [support absolute filepath in name argument](https://www.riverbankcomputing.com/pipermail/pyqt/2022-June/044700.html)

some patches are not required, but useful for debugging,
for example to dump the generated cmake files to stdout

<details>
<summary>
Configure summary
</summary>

these are the default settings. i have not tried to build all modules

```
---- Configure summary ----

These bindings will be built:

  [x] QtCore
  [x] QtNetwork
  [x] QtGui
  [x] QtQml
  [x] QtWidgets
  [x] QtDBus
  [ ] QtDesigner
  [ ] QtHelp
  [x] QtOpenGL
  [x] QtOpenGLWidgets
  [x] QtPrintSupport
  [x] QtQuick
  [ ] QtQuick3D
  [x] QtQuickWidgets
  [x] QtSql
  [x] QtSvg
  [x] QtSvgWidgets
  [x] QtTest
  [x] QtXml
  [ ] QtMultimedia
  [ ] QtMultimediaWidgets
  [ ] QtPositioning
  [ ] QtRemoteObjects
  [ ] QtSensors
  [ ] QtSerialPort
  [x] QtWebChannel
  [x] QtWebSockets
  [ ] QtBluetooth
  [ ] QtNfc
  [ ] QAxContainer
```

</details>

<details>
<summary>
pyqt6-hello-world.py
</summary>

```py
import sys

#from PyQt6.QtCore import *
#from PyQt6.QtGui import *
from PyQt6.QtWidgets import *

app = QApplication(sys.argv)

w = QWidget()
b = QLabel(w)
b.setText("hello")
w.setGeometry(100, 100, 200, 50)
b.move(50, 20)
w.setWindowTitle("pyqt")
w.show()

sys.exit(app.exec())

#app.exec(); del app
```

</details>



<details>
<summary>
fixed: runtime error: Cannot add multiple registrations for QtQml
</summary>

the qtdeclarative bindings fail in checkPhase / on runtime

```sh
$ python3 -c "import PyQt6.QtQml"
Cannot add multiple registrations for QtQml
Aborted (core dumped)
```

[qtdeclarative/src/qml/qml/qqmlmetatype.cpp](https://code.qt.io/cgit/qt/qtdeclarative.git/tree/src/qml/qml/qqmlmetatype.cpp?h=dev#n272)

```cc
void QQmlMetaType::qmlInsertModuleRegistration(const QString &uri, void (*registerFunction)())
{
    QQmlMetaTypeDataPtr data;
    if (data->moduleTypeRegistrationFunctions.contains(uri))
        qFatal("Cannot add multiple registrations for %s", qPrintable(uri));
    else
        data->moduleTypeRegistrationFunctions.insert(uri, registerFunction);
}
```

im clueless for now. hope to fix this soon
first guess: this needs proper qt env with qml import paths

<details>
<summary>
gdb backtrace
</summary>

```
$ gdb python3

(gdb) run ./result/bin/.pyqt6-test.py-wrapped
Starting program: /nix/store/vzqny68wq33dcg4hkdala51n5vqhpnwc-python3-3.9.12/bin/python3.9 ./result/bin/.pyqt6-test.py-wrapped
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/fz33c1mfi2krpg1lwzizfw28kj705yg0-glibc-2.34-210/lib/libthread_db.so.1".
trying to import PyQt6.QtQml
warning: File "/nix/store/g9x3cm8wdmpm86pyiff6arx8skmdwkn7-gcc-11.3.0-lib/lib/libstdc++.so.6.0.29-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load:/nix/store/8dn12i3d7harw8g7dzk6dy7c5diz5ibp-gcc-11.3.0-lib".
Cannot add multiple registrations for QtQml

Program received signal SIGABRT, Aborted.
0x00007ffff7934c1f in __pthread_kill_implementation () from /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libc.so.6

(gdb) bt
#0  0x00007ffff7934c1f in __pthread_kill_implementation () from /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libc.so.6
#1  0x00007ffff78ea042 in raise () from /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libc.so.6
#2  0x00007ffff78d549c in abort () from /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libc.so.6
#3  0x00007fffe8878520 in qAbort() () from /nix/store/mpxx1jy7bwvg4vd98h4p2h2d3zvq4blx-qtbase-6.3.0/lib/libQt6Core.so.6
#4  0x00007fffe8878c70 in QMessageLogger::fatal(char const*, ...) const () from /nix/store/mpxx1jy7bwvg4vd98h4p2h2d3zvq4blx-qtbase-6.3.0/lib/libQt6Core.so.6

// Cannot add multiple registrations for QtQml

#5  0x00007fffe905f8ae in QQmlMetaType::qmlInsertModuleRegistration(QString const&, void (*)()) [clone .cold] ()
   from /nix/store/s6z72847bbdf6xgacb17455j44xrnkl3-qtdeclarative-6.3.0/lib/libQt6Qml.so.6

#6  0x00007fffe9e310de in _GLOBAL__sub_I_qtqml_qmltyperegistrations.cpp ()
   from /nix/store/p4sazzaa430i7lxklb0ghnx8syxrz25c-python3.9-PyQt6-6.3.0/lib/python3.9/site-packages/PyQt6/QtQml.abi3.so

#7  0x00007ffff7fcef8e in call_init () from /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/ld-linux-x86-64.so.2
```

</details>

&rarr; maybe a bug in `_GLOBAL__sub_I_qtqml_qmltyperegistrations.cpp`

this cmake output looks suspicious
especially `Automatic QML type registration for target QtQml`
produced by [qtdeclarative/src/qml/Qt6QmlMacros.cmake](https://code.qt.io/cgit/qt/qtdeclarative.git/tree/src/qml/Qt6QmlMacros.cmake#n2362)

```
[ 14%] Running moc --collect-json for target QtQml
[ 14%] Running rcc for resource qmake_QtQml
[ 14%] Automatic QML type registration for target QtQml
Consolidate compiler generated dependencies of target QtQml
[ 15%] Building CXX object QtQml/CMakeFiles/QtQml.dir/QtQml_autogen/mocs_compilation.cpp.o
```

&rarr; disable cmake's qml type generator

```cmake
qt_add_qml_module(
    URI "..."
    VERSION "..."
    NO_GENERATE_QMLTYPES
)
```

fixed : )

now

```py
#!/usr/bin/env python3
print("trying to import PyQt6.QtQml")
import PyQt6.QtQml
print("PyQt6.QtQml", PyQt6.QtQml)
```

prints

```
trying to import PyQt6.QtQml
PyQt6.QtQml <module 'PyQt6.QtQml' from '/nix/store/05q4krlqix8388ddz0sifd81hb51008m-python3.9-PyQt6-6.3.0/lib/python3.9/site-packages/PyQt6/QtQml.abi3.so'>
```

</details>

**TODO**

sync with #178171 (qt 6.3.1)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
python3Packages.pyqt6: init at 6.3.1

Commit:
python3Packages.qmake2cmake: init at unstable-2022-07-06

nrdxp commented on pkgs/development/python-modules/pyqt-builder/0005-nix-use-NIX_BUILD_CORES-to-build-faster-by-default.patch:
```diff
@@ -0,0 +1,28 @@
++        default_jobs = os.environ.get("NIX_BUILD_CORES", 1)
+         options.append(
+                 Option('jobs', option_type=int,
++                        default=default_jobs,
```
should we make this conditional on whether `enableParallelBuilding` is set?

milahu replied:
yes

https://github.com/NixOS/nixpkgs/blob/9585acba1c3d8c3bcbd53f0a48d884d3070659aa/pkgs/stdenv/generic/setup.sh#L1078

milahu commented:
> pythonPackages.sip: add patch: feat: check and generate bindings in parallel

parallel build is broken

pyqt declares the build order in `PyQt6-6.3.0/project.py`

```py
        # Each set of bindings must appear after any set they depend on.
        self.bindings_factories = [QtCore, QtNetwork, QtGui, QtQml, QtWidgets,
                QtDBus, QtDesigner, QtHelp, QtOpenGL, QtOpenGLWidgets,
                # ...
```

but cmake builds the main targets in parallel, which throws random errors like

```
[ 19%] Building CXX object QtDBus/CMakeFiles/QtDBus.dir/sipQtDBusQDBusReply0100QStringList.cpp.o
/build/tmpr44l0mn6/QtOpenGL/sipQtOpenGLQOpenGLFunctions_ES2.cpp: In function ‘PyObject* meth_QOpenGLFunctions_ES2_initializeOpenGLFunctions(PyObject*, PyObject*)’:
/build/tmpr44l0mn6/QtOpenGL/sipQtOpenGLQOpenGLFunctions_ES2.cpp:40:12: error: ‘::QOpenGLFunctions_ES2’ has not been declared; did you mean ‘sipType_QOpenGLFunctions_ES2’?
    40 |          ::QOpenGLFunctions_ES2 *sipCpp;
       |            ^~~~~~~~~~~~~~~~~~~~
       |            sipType_QOpenGLFunctions_ES2
```

&rarr; missing feature in qmake2cmake

draft until this is stable

milahu commented:
fixed parallel configure
with 32 cores, the build now takes 400 seconds instead of 500 seconds = 20% faster

codegen must run serial
QtCore codegen must run before QtOpenGL codegen
because QtCore defines `%Feature PyQt_OpenGL_ES2`

nrdxp commented:
Builds fine for me, good work.

milahu commented:
[PyQt6-WebEngine](https://pypi.org/project/PyQt6-WebEngine/) requires qt 6.3.1 &rarr; #178171

Commit:
python3Packages.pyqt6-webengine: init at 6.3.1

nrdxp commented:
Well I had a working package for 6.3.0 using my [qmake hack](https://github.com/nrdxp/nixpkgs/blob/pyqt6-new/pkgs/development/python-modules/pyqtwebengine/6.x.nix). I'm typing from a qutebrowser build making use of it in fact.

I played around with your branch yesterday to see if I could maybe get a dependant PR going, but no luck so far.

milahu commented:
aah im blind. of course there is 6.3.0 in https://pypi.org/project/PyQt6-WebEngine/#history

can you try ...?

<details>
<summary>
python3Packages.pyqt6-webengine-init-at-6.3.0.patch
</summary>

```diff
From 4c322267dae8906e962c97272101c53e7ca9c807 Mon Sep 17 00:00:00 2001
From: Milan Hauth <milahu@gmail.com>
Date: Wed, 22 Jun 2022 10:31:32 +0200
Subject: [PATCH] DRAFT python3Packages.pyqt6-webengine: init at 6.3.0

---
 .../python-modules/pyqtwebengine/6.x.nix      | 80 +++++++++++++++++++
 pkgs/top-level/python-packages.nix            |  4 +
 2 files changed, 84 insertions(+)
 create mode 100644 pkgs/development/python-modules/pyqtwebengine/6.x.nix

diff --git a/pkgs/development/python-modules/pyqtwebengine/6.x.nix b/pkgs/development/python-modules/pyqtwebengine/6.x.nix
new file mode 100644
index 00000000000..478b7218b4c
--- /dev/null
+++ b/pkgs/development/python-modules/pyqtwebengine/6.x.nix
@@ -0,0 +1,80 @@
+# based on pkgs/development/python-modules/pyqt/6.x.nix
+
+{ lib
+, pythonPackages
+, pkg-config
+, cmake
+, qtbase
+, qtwebengine
+, wrapQtAppsHook
+}:
+
+let
+  inherit (pythonPackages) buildPythonPackage python isPy27 pyqt6 enum34 sip pyqt-builder qmake2cmake;
+in buildPythonPackage rec {
+  pname = "PyQt6-WebEngine";
+  version = "6.3.0";
+  format = "pyproject";
+
+  disabled = isPy27;
+
+  src = pythonPackages.fetchPypi {
+    pname = "PyQt6_WebEngine";
+    inherit version;
+    sha256 = "sha256-qyrtvuxU8bz/hy99/CNqoPzktVzTD2CMqJtAjunoFHs=";
+  };
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace "[tool.sip.project]" "[tool.sip.project]''\nsip-include-dirs = [\"${pyqt6}/${python.sitePackages}/pyqt6/bindings\"]"
+  '';
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    pkg-config
+    qmake2cmake
+    cmake
+    sip
+    qtbase
+    qtwebengine
+    pyqt-builder
+  ];
+
+  buildInputs = [
+    sip
+    qtbase
+    qtwebengine
+  ];
+
+  propagatedBuildInputs = [ pyqt6 ];
+
+  dontWrapQtApps = true;
+
+  # Checked using pythonImportsCheck
+  doCheck = false;
+
+  pythonImportsCheck = [
+    "PyQt6.QtWebEngineCore"
+    "PyQt6.QtWebEngineWidgets"
+  ];
+
+  # help sip to import /build/PyQt6-6.3.0/project.py
+  # to enable parallel configure + codegen
+  enableParallelBuilding = true;
+  configurePhase = ''
+    export PYTHONPATH="$PWD:$PYTHONPATH"
+  '';
+
+  passthru = {
+    inherit wrapQtAppsHook;
+  };
+
+  meta = with lib; {
+    description = "Python bindings for Qt6 WebEngine";
+    homepage    = "http://www.riverbankcomputing.co.uk";
+    license     = licenses.gpl3;
+    platforms   = lib.lists.intersectLists qtwebengine.meta.platforms platforms.mesaPlatforms;
+    maintainers = with maintainers; [ nrdxp milahu ];
+  };
+}
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index d12435b5d42..03c5ed1d397 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -7855,6 +7855,10 @@ in {
     pythonPackages = self;
   };
 
+  pyqt6-webengine = pkgs.qt6Packages.callPackage ../development/python-modules/pyqtwebengine/6.x.nix {
+    pythonPackages = self;
+  };
+
   pyquery = callPackage ../development/python-modules/pyquery { };
 
   pyquaternion = callPackage ../development/python-modules/pyquaternion { };
-- 
2.36.1

```

</details>

milahu commented:
UnboundLocalError was a bug in my patch for sip. fixed in my last force-push



now make cant find a qt header

```
  [ 65%] Building CXX object QtWebEngineCore/CMakeFiles/QtWebEngineCore.dir/sipQtWebEngineCoreQWebEngineSettings.cpp.o
  /build/PyQt6_WebEngine-6.3.0/sip/QtWebEngineCore/qwebenginecontextmenurequest.sip:26:10: fatal error: qwebenginecontextmenurequest.h: No such file or directory
     26 | #include <qwebenginecontextmenurequest.h>
```

the file is in
/nix/store/*-qtwebengine-6.3.0-dev/include/QtWebEngineCore/qwebenginecontextmenurequest.h

looks like a bug in qmake2cmake

```
  ./CMakeLists.txt:110 find_package(QT NAMES Qt5 Qt6 REQUIRED)
  ./CMakeLists.txt:111 find_package(Qt${QT_VERSION_MAJOR} REQUIRED COMPONENTS Core Gui WebEngineQuick WebEngineWidgets)
```

probably find_package is missing the component WebEngineCore

see my branch [pyqt6-cmake-webengine](https://github.com/milahu/nixpkgs/tree/pyqt6-cmake-webengine)

will fix this later today ...

milahu commented:
fixed in https://github.com/milahu/qmake2cmake/commit/edfcd13e4c84017634a48debcb8ef1b04a86c760 and https://github.com/milahu/qmake2cmake/commit/611a3738860b1997d279f077814b3c59949f50fb

Commit:
python3Packages.pyqt6-builder: init at 1.12.2

qt6 qmake is broken in nixpkgs
instead of qmake, use qmake2cmake + cmake

keep python3Packages.pyqt-builder for pyqt5

Commit:
python3Packages.sip-pyqt6: init at 6.6.1

Commit:
qutebrowser-qt6: init at unstable-2022-06-22

nrdxp commented:
Very nice! It is working now as well. I did have to add this small patch to get it to run under wayland:
```diff
diff --git a/pkgs/development/libraries/qt-6/default.nix b/pkgs/development/libraries/qt-6/default.nix
index 5dbb241c6c5..978e9e528ae 100644
--- a/pkgs/development/libraries/qt-6/default.nix
+++ b/pkgs/development/libraries/qt-6/default.nix
@@ -84,7 +84,7 @@ let
       qtwebview = callPackage ./modules/qtwebview.nix { };
 
       wrapQtAppsHook = makeSetupHook {
-          deps = [ makeWrapper ];
+          deps = [ makeWrapper ] ++ lib.optional stdenv.isLinux self.qtwayland.dev;
         } ./hooks/wrap-qt-apps-hook.sh;
     };
 
```

rnhmjoj commented:
This is not a criticism, I'm just surprised: is porting the previous qmake builder to qt6 so bad that we rather have 7 patches in this software? Also, is PyQtBuilder expected to move to cmake and could these patches be upstreamed?

nrdxp commented:
@rnhmjoj, I did try to find those answers when I first started by original PR which this one supercedes, but I couldn't find anything conclusive. I would assume that if qmake is deprecated that they would eventually want to follow suit with the rest of the qt stack and move to cmake, but I couldn't find any word on that, or any work toward it in their development builds.

Perhaps they would accept the changes as a patch though :shrug:

milahu commented:
> is porting the previous qmake builder to qt6 so bad that we rather have 7 patches in this software?

fixing qmake would be the right thing to do
but since qmake is deprecated in qt6, i focussed on cmake and qmake2cmake

qmake would need patches for nix multi-prefix filesystem layout
similar to the cmake patches for qt6

with pyside6, pyqt6 is more-or-less legacy software
so i would not expect big changes

milahu commented:
> I did have to add this small patch to get it to run under wayland:

qtwayland was removed from wrapQtAppsHook in #174946

rnhmjoj commented:
> with pyside6, pyqt6 is more-or-less legacy software

I see, unfortunately it [seems](https://github.com/qutebrowser/qutebrowser/issues/3625#issuecomment-1100623372) pyside is quite there yet.

nrdxp commented:
> → added qtwayland to buildInputs of pyqt6 and pyqt6-webengine

Unfortunately that doesn't seem to work:
```
16:00:40 WARNING: Could not find the Qt platform plugin "wayland" in ""
```

milahu commented:
add qtbase + qtwayland to buildInputs / propagatedBuildInputs of qutebrowser? (etc)

nrdxp commented:
> add qtbase + qtwayland to buildInputs / propagatedBuildInputs of qutebrowser? (etc)

still doesn't work unfortunately

milahu commented:
sounds like a bug in the qt setup-hook.sh ...

add qtbase + qtwayland to buildInputs and nativeBuildInputs of qutebrowser? (sorry for the chit chat)

nrdxp commented on pkgs/development/python-modules/qmake2cmake/readme.md:
```diff
@@ -0,0 +1,21 @@
+  cmakeFlags = [ "--debug-output" ];
+  makeFlags = [ "-d" ];
+  # debug qmake2cmake
+  QMAKE2CMAKE_DEBUG_DUMP_FILES = "1"; # *.pro CMakeLists.txt (TODO: cmake_install.cmake)
```
should we comment this out before merging?

milahu replied:
thats just the readme

SuperSandro2000 replied:
I don't understand this or the comments below. What is it supposed to tell me?

milahu replied:
with these envs, generated files are dumped to stdout

this is useful when i dont have root access to the builder, and cannot inspect files in the build sandbox

nrdxp reviewed (approved):
All the important pieces are there and functioning. Just a few little nits; mainly I'm wondering if we should revert https://github.com/NixOS/nixpkgs/pull/174946#issuecomment-1174422158?

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7848,6 +7858,10 @@ in {
     pythonPackages = self;
   };
 
+  pyqt6-webengine = pkgs.qt6Packages.callPackage ../development/python-modules/pyqtwebengine/6.x.nix {
```
```suggestion
  pyqt6-webengine = pkgs.qt6Packages.callPackage ../development/python-modules/pyqtwebengine/6.nix {
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7823,12 +7823,22 @@ in {
 
+  pyqt6-builder = callPackage ../development/python-modules/pyqt-builder/6.x.nix { };
+
   pyqt4 = callPackage ../development/python-modules/pyqt/4.x.nix { };
 
   pyqt5 = callPackage ../development/python-modules/pyqt/5.x.nix { };
 
+  pyqt6 = callPackage ../development/python-modules/pyqt/6.x.nix { };
```
```suggestion
  pyqt4 = callPackage ../development/python-modules/pyqt/4.nix { };

  pyqt5 = callPackage ../development/python-modules/pyqt/5.nix { };

  pyqt6 = callPackage ../development/python-modules/pyqt/6.nix { };
```

I know this is a bit unrelated but the shorter name without x is nicer.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29327,6 +29327,8 @@ with pkgs;
 
   qutebrowser = libsForQt5.callPackage ../applications/networking/browsers/qutebrowser { };
 
+  qutebrowser-qt6 = qt6Packages.callPackage ../applications/networking/browsers/qutebrowser/qutebrowser-qt6.nix { };
```
```suggestion
  qutebrowser-qt6 = qt6Packages.callPackage ../applications/networking/browsers/qutebrowser/qt6.nix { };
```

SuperSandro2000 commented on pkgs/development/python-modules/sip/sip-pyqt6.nix:
```diff
@@ -0,0 +1,39 @@
+  patches = old.patches ++ [
+    # cosmetic
+    ./fix-unbuffer-make-output.patch
+    # cosmetic
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/sip/sip-pyqt6.nix:
```diff
@@ -0,0 +1,39 @@
+  # https://www.riverbankcomputing.com/pipermail/pyqt/2022-June/044700.html
+  # required to build pyqt6
+  postPatch = ''
+    sed -i.bak -E \
```
```suggestion
    sed -i -E \
```

SuperSandro2000 commented on pkgs/development/python-modules/sip/feat-parallel-configure.patch:
```diff
@@ -0,0 +1,275 @@
+From 3b86b3489dcbdcef3a83312cd168e05458a0d0ad Mon Sep 17 00:00:00 2001
```
Can we try to upstream this or at least parts?

milahu replied:
[upstream says no](https://www.riverbankcomputing.com/pipermail/pyqt/2022-June/044717.html)

> Thanks but I’m not looking at optimisations until all the C code has 
> been replaced by Python code.
> 
> Phil

SuperSandro2000 commented on pkgs/development/python-modules/qmake2cmake/default.nix:
```diff
@@ -0,0 +1,56 @@
+  ];
+
+  checkInputs = [
+    pytest
```
```suggestion
    pytestCHeckHook
```

SuperSandro2000 commented on pkgs/development/python-modules/qmake2cmake/default.nix:
```diff
@@ -0,0 +1,56 @@
+    sha256 = "sha256-8J+5U+hyAy7Po2eGpaq9qEe8omgRxv5dpUNFFq+YbLo=";
+  };
+
+  /*
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-5h5891RDGgisv5VJUBSepbZq/3c4CnsPdTk2iJ3rKmg=";
+  };
+  */
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/qmake2cmake/default.nix:
```diff
@@ -0,0 +1,56 @@
+  pname = "qmake2cmake";
+  version = "1.0.1-unstable-2022-06-24";
+
+  src = fetchFromGitHub {
+    # https://github.com/milahu/qmake2cmake/tree/fix-for-pyqt-builder
```
We should also add a comment why we need the fork.

SuperSandro2000 commented on pkgs/development/python-modules/qmake2cmake/default.nix:
```diff
@@ -0,0 +1,56 @@
+
+buildPythonPackage rec {
+  pname = "qmake2cmake";
+  version = "1.0.1-unstable-2022-06-24";
```
```suggestion
  version = "unstable-2022-06-24";
```

SuperSandro2000 commented on pkgs/development/python-modules/qmake2cmake/default.nix:
```diff
@@ -0,0 +1,56 @@
+/*
+nix-shell -E 'with import <nixpkgs> { }; python3Packages.callPackage ./default.nix { }'
+*/
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+
+  meta = with lib; {
+    description = "Python bindings for Qt6 WebEngine";
+    homepage    = "http://www.riverbankcomputing.co.uk";
```
```suggestion
    homepage    = "https://riverbankcomputing.com/";
```

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+    "PyQt6.QtWebEngineWidgets"
+  ];
+
+  # help sip to import /build/PyQt6-6.3.0/project.py
+  # to enable parallel configure + codegen
```
```suggestion
```

milahu replied:
lets keep this, as its not obvious
maybe rephrase to

> needed for sip-pyqt6 with feat-parallel-configure.patch

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+  # help sip to import /build/PyQt6-6.3.0/project.py
+  # to enable parallel configure + codegen
+  enableParallelBuilding = true;
+  configurePhase = ''
```
```suggestion

  # help sip to import /build/PyQt6-6.3.0/project.py
  preConfigure = ''
```

milahu replied:
need to test this

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+
+  dontWrapQtApps = true;
+
+  # Checked using pythonImportsCheck
```
```suggestion
```
pythonImportsCheck are not a replacement for a proper test suite.

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+  };
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+    --replace \
+    '[tool.sip.project]' \
+    '[tool.sip.project]
+    sip-include-dirs = ["${pyqt6}/${python.sitePackages}/PyQt6/bindings"]'
```
```suggestion
    substituteInPlace pyproject.toml \
      --replace '[tool.sip.project]' '[tool.sip.project]\nsip-include-dirs = ["${pyqt6}/${python.sitePackages}/PyQt6/bindings"]'
```

milahu replied:
this throws

> ```pip._vendor.tomli.TOMLDecodeError: Expected newline or end of document after a statement (at line 23, column 19)```

line 23 is

> ```[tool.sip.project]```

fix:

```diff
-  '...\n...'
+  $'...\n...'
```

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+
+  disabled = isPy27;
+
+  src = pythonPackages.fetchPypi {
```
```suggestion
  src = fetchPypi {
```

SuperSandro2000 commented on pkgs/development/python-modules/pyqtwebengine/6.x.nix:
```diff
@@ -0,0 +1,86 @@
+
+# based on pkgs/development/python-modules/pyqt/6.x.nix
+
+let
+  inherit (pythonPackages) buildPythonPackage python isPy27 pyqt6 enum34 sip-pyqt6 pyqt6-builder qmake2cmake;
+in buildPythonPackage rec {
```
```suggestion
with pythonPackages;

buildPythonPackage rec {
```

now we don't need to maintain that list.

SuperSandro2000 commented on pkgs/development/python-modules/pyqt/6.x.nix:
```diff
@@ -0,0 +1,114 @@
+    WebSocketsEnabled = withWebSockets;
+  };
+
+  # help sip to import /build/PyQt6-6.3.0/project.py
+  # to enable parallel configure + codegen
+  enableParallelBuilding = true;
+  configurePhase = ''
```
```suggestion
  enableParallelBuilding = true;

  # help sip to import /build/PyQt6-6.3.0/project.py
  preConfigure = ''
```

SuperSandro2000 commented on pkgs/applications/networking/browsers/qutebrowser/qutebrowser-qt6.nix:
```diff
@@ -0,0 +1,131 @@
+    stripRoot = false;
+  };
+
+in python3Packages.buildPythonApplication rec {
```
```suggestion
in python3.pkgs.buildPythonApplication rec {
```

SuperSandro2000 commented on pkgs/applications/networking/browsers/qutebrowser/qutebrowser-qt6.nix:
```diff
@@ -0,0 +1,131 @@
+    description = "Keyboard-focused browser with a minimal GUI";
+    license     = licenses.gpl3Plus;
+    maintainers = with maintainers; [ jagajaga rnhmjoj ebzzry dotlambda ];
+    inherit (python3Packages.pyqt6-webengine.meta) platforms;
```
```suggestion
    inherit (python3.pkgs.pyqt6-webengine.meta) platforms;
```

SuperSandro2000 commented on pkgs/applications/networking/browsers/qutebrowser/qutebrowser-qt6.nix:
```diff
@@ -0,0 +1,131 @@
+    docbook_xml_dtd_45 docbook_xsl libxml2 libxslt
+  ];
+
+  propagatedBuildInputs = with python3Packages; ([
```
```suggestion
  propagatedBuildInputs = with python3.pkgs; ([
```

SuperSandro2000 reviewed (commented):
Mostly stylistic comments. Maybe we should split some of the qt python packages into their own PRs to move this faster forward.

Ordering is a bit strange in some packages but I didn't comment that.

milahu commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7823,12 +7823,22 @@ in {
 
+  pyqt6 = callPackage ../development/python-modules/pyqt/6.x.nix { };
+
+  pyqt6_with_qtmultimedia = self.pyqt6.override {
+    withMultimedia = true;
+  };
+
```
same for pyqt5_with_qtmultimedia ...

nrdxp commented:
qutebrowser no longer builds after the last commit:
```
error: Function called without required argument "pytestCHeckHook" at /nix/store/4a10i0gh644haidp2sswr9rwci2hh4jb-source/pkgs/development/python-modules/qmake2cmake/default.nix:10, did you mean "pytestCheckHook"?
```

nrdxp commented:
now `qmake2cmake-unstable` is failing during the test phase it seems. Here's the summary at the end of the failure:
```
       > =========================== short test summary info ============================
       > FAILED tests/test_conversion.py::test_subdirs - AssertionError: assert ['CMak...
       > FAILED tests/test_conversion.py::test_qml_modules - AssertionError: assert '\...
       > FAILED tests/test_conversion.py::test_install_targets - assert '\n# install_l...
       > ======================== 3 failed, 95 passed in 22.06s =========================
```

nrdxp commented:
Oh also, now that qt 6.3.1 is merged, we should probably update this to match otherwise it may not work after being merged.

milahu commented:
should be fixed ... needs some polish, drafted, feel free to help

Commit:
python3Packages.pyqt5: mv 5.x.nix 5.nix

Commit:
python3Packages.pyqt4: mv 4.x.nix 4.nix

milahu commented:
mkay. now we just gotta fix qutebrowser-qt6

milahu commented:
> getting weird parse errors

good news, that was the debug output

bug was in the output path, so the CMakeLists.txt files were in the wrong location
so cmake said this for all subdirs: 

```
  CMake Error at CMakeLists.txt:84 (add_subdirectory):
    The source directory

      /build/tmpa2fyk8aw/QtWebSockets

    does not contain a CMakeLists.txt file.
```

fixed

next bug: cmake is trying to run `sip-distinfo`

> ```/nix/store/bwlp4fdf5gldsq5iw48i419c1dp23yca-python3.10-pip-22.1.2/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/sip-distinfo```

which does not exist ... fixed

next bug:

```
Executing pipInstallPhase
/build/PyQt6-6.3.1/dist /build/PyQt6-6.3.1
Processing ./PyQt6-6.3.1-cp37-abi3-manylinux1_x86_64.whl
ERROR: Could not find a version that satisfies the requirement PyQt6-sip<14,>=13.4 (from pyqt6) (from versions: none)
ERROR: No matching distribution found for PyQt6-sip<14,>=13.4
```

&rarr; update pyqt-sip

next bug: `internal compiler error: Segmentation fault` ...
this looks like overloading (more jobs than NIX_BUILD_CORES)
todo: analyze with nix-build-profiler https://github.com/NixOS/nixpkgs/pull/180894


latest version is in my [pyqt6-cmake-test](https://github.com/milahu/nixpkgs/tree/pyqt6-cmake-test) branch

Commit:
qt6.qtdeclarative: add patch: fix qt6_add_qml_module

milahu commented:
@NickCao this looks like a bug in your patch for cmake in qt6

/nix/store/4y3ag0gkg382is1rkzldr8b379wpfp74-qtdeclarative-6.3.1-dev/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake

function(_qt_internal_scan_qml_imports target imports_file_var when_to_scan)

```
_qt_additional_packages_prefix_paths =
...
/nix/store/4y3ag0gkg382is1rkzldr8b379wpfp74-qtdeclarative-6.3.1-dev/lib/cmake
/nix/store/j4ik6hsadspyc27qxjh7j9s5rfn8dyri-qtdeclarative-6.3.1/lib/cmake
...
```

but these paths should not have the `/lib/cmake` suffix

also, this foreach loop is broken:

```cmake
# /nix/store/4y3ag0gkg382is1rkzldr8b379wpfp74-qtdeclarative-6.3.1-dev/lib/cmake/Qt6Qml/Qt6QmlMacros.cmake

foreach(root IN ITEMS "/nix/store/j4ik6hsadspyc27qxjh7j9s5rfn8dyri-qtdeclarative-6.3.1;${_qt_additional_packages_prefix_paths}")
```

this gives only one iteration with `root = /nix/store/j4ik6hsadspyc27qxjh7j9s5rfn8dyri-qtdeclarative-6.3.1;/nix/store/i6svmmq24wl4i10dx0m8blrpl3m3fq4b-python3-3.10.5/lib/cmake;/nix/store/rnrhx246in8a9gzd3xlg8hk0sj1fjxwf-hook/lib/cmake;/nix/store/9yr1ay52lrf8v5ah84xxhwm8dan1x5qf-hook/lib/cmake;........`

fix: remove doublequotes
cleanup: remove `IN ITEMS`

```cmake
foreach(root /nix/store/j4ik6hsadspyc27qxjh7j9s5rfn8dyri-qtdeclarative-6.3.1;${_qt_additional_packages_prefix_paths})
```

milahu commented:
> sounds like a bug in the qt setup-hook.sh ...

2 bugs, fixed in #181465

this should work now

i also tested qutebrowser-qt6 on wayland (plasma desktop)

milahu commented:
abandoned. sorry guys, i have no more time for this

nrdxp commented:
> abandoned. sorry guys, i have no more time for this

Thanks for all your effort here! I've been short on time as well but I'll see if I can maybe inch this toward the finish line soon :crossed_fingers:

milahu commented:
i guess the easiest path is to patch qmake to accept symlinks, and use symlinkJoin to merge all qt inputs

the qmake2cmake route is unstable, as qmake2cmake has bugs and fails in edge-cases
